instance/models/appserver.py

```diff
@@ -0,0 +1,254 @@
+    )
+    _status = models.CharField(
+        max_length=20,
+        default=status.default_state_class.state_id,
+        choices=status.model_field_choices,
+        db_index=True,
+        db_column='status',
+    )
+    # State transitions:
+    _status_to_waiting_for_server = status.transition(
+        from_states=Status.New, to_state=Status.WaitingForServer
+    )
+    _status_to_configuring_server = status.transition(
+        from_states=Status.WaitingForServer, to_state=Status.ConfiguringServer
+    )
+    _status_to_error = status.transition(
+        from_states=Status.WaitingForServer, to_state=Status.Error
+    )
+    _status_to_running = status.transition(
+        from_states=Status.ConfiguringServer, to_state=Status.Running
+    )
+    _status_to_configuration_failed = status.transition(
+        from_states=Status.ConfiguringServer, to_state=Status.ConfigurationFailed
+    )
+    _status_to_terminated = status.transition(
+        from_states=Status.Running, to_state=Status.Terminated
+    )
+
+    name = models.CharField(max_length=250, blank=False)
+    server = models.OneToOneField(OpenStackServer, on_delete=models.CASCADE, related_name='+')
+    # The Instance that owns this. Instance will get related_name accessors like 'openedxappserver_set'
+    owner = models.ForeignKey(InstanceReference, on_delete=models.CASCADE, related_name='%(class)s_set')
+
+    class Meta:
+        abstract = True
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.logger = AppServerLoggerAdapter(logger, {'obj': self})
+
+    def __str__(self):
+        return self.name
+
+    @property
+    def instance(self):
+        """
+        Get the Instance that owns this AppServer
+        """
+        return self.owner.instance
+
+    @property
+    def event_context(self):
+        """
+        Context dictionary to include in events
+        """
+        context = self.instance.event_context  # dict with instance_id
+        context.update({'appserver_id': self.pk, 'appserver_type': self.__class__.__name__})
+        return context
+
+    def set_field_defaults(self):
+        """
+        Set default values.
+
+        We don't use django field 'defaults=callable' because changes to the field defaults
+        affect automatically generated migrations, generating a new one when settings don't match
+
+        This method is called by clean_fields() before validation and save()
+        """
+        pass
+
+    def clean_fields(self, exclude=None):
+        """
+        Set default values before validation.
+        """
+        if not self.pk:
+            self.set_field_defaults()
+        super().clean_fields(exclude=exclude)
+
+    def save(self, **kwargs):
+        if self.pk:
+            # We are changing an existing instance. But most AppServer fields are meant to be
+            # immutable. Only 'status' can change.
```

Comment:
@bradenmacdonald "We are changing an existing app server", perhaps? Also, based on the code below "modified" can change as well, not just "status".